### PR TITLE
Amélioration des performances de l'export

### PR DESCRIPTION
**Pull request by yohanboniface**
- split into one file per resource
- cache some foreignkeys (user, session, municipality…)
- add some JOIN
- distribute over processes
- better slice querysets when using helpers.batch
- add Resource.as_export for filtering/customization of exported fields

### CODE REVIEW

File: ban/db/cache.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+"""
+Very simple custom cache.
+functools.lru_cache does not make it, as it only works with hashable types,
+while we need peewee.Expression to be memoized.
+On the other hand, we want to select witch SQL queries will be cached (i.e. not
```
Comment (davidbgk) [inline ban/db/cache.py]: Beware of witches 😈  `which` 😉

File: ban/db/fields.py [hunk 2]
```diff
@@ -104,11 +105,30 @@ def contains(self, dt):
...
 class ForeignKeyField(peewee.ForeignKeyField):
 
     __data_type__ = int
     __schema_type__ = 'integer'
 
+    def __init__(self, *args, **kwargs):
+        self.cache = kwargs.pop('cache', False)
+        super().__init__(*args, **kwargs)
+
+    def _get_descriptor(self):
```
Comment (davidbgk) [inline ban/db/fields.py]: Requires a comment to explain why we are adding it.

File: ban/commands/export.py [hunk 3]
```diff
@@ -12,10 +27,29 @@ def resources(path, **kwargs):
...
+            print('Exporting to', f.name)
+            query = query.order_by(resource.pk)
+            results = []
+            for result in helpers.batch(process_resource, query,
+                                        chunksize=1000, total=query.count()):
+                results.append(result)
+                if len(results) == 10000:
+                    f.write('\n'.join(results) + '\n')
+                    f.flush()
+                    os.fsync(f.fileno())
```
Comment (davidbgk) [inline ban/commands/export.py]: Is that still useful now? I remember it wasn't working as expected.

File: ban/commands/export.py [hunk 4]
```diff
@@ -12,10 +27,29 @@ def resources(path, **kwargs):
...
+                    results = []
+            if results:
+                f.write('\n'.join(results))
```
Comment (davidbgk) [inline ban/commands/export.py]: I think we should add a new line at the end of the file to be consistent with the case when there are no remaining results.

File: ban/commands/export.py [hunk 5]
```diff
@@ -12,10 +27,29 @@ def resources(path, **kwargs):
...
+    resources = [Municipality, PostCode, Group, HouseNumber]
+    for resource in resources:
+        query = QUERIES.get(resource)
+        filename = '{}.ndjson'.format(resource.__name__.lower())
+        with Path(path).joinpath(filename).open(mode='w') as f:
+            print('Exporting to', f.name)
+            query = query.order_by(resource.pk)
+            results = []
+            for result in helpers.batch(process_resource, query,
+                                        chunksize=1000, total=query.count()):
```
Comment (davidbgk) [inline ban/commands/export.py]: No var for `chunksize` nor `10000` below?

File: ban/commands/helpers.py [hunk 6]
```diff
@@ -69,11 +69,16 @@ def collect_report(func, chunk):
...
-    return len(results), reports
+    return results, reports
 
 
 class ChunkedPool(Pool):
 
+    @classmethod
+    def _get_tasks_from_query(cls, func, query, chunksize):
+        for idx in range(0, query.count(), chunksize):
+            yield (func, list(query.limit(chunksize).offset(idx)))
```
Comment (davidbgk) [inline ban/commands/helpers.py]: 👍

File: ban/commands/init.py [hunk 7]
```diff
@@ -35,7 +35,7 @@ def init(*paths, limit=0, **kwargs):
             print('Computing file size')
             total = sum(1 for line in helpers.iter_file(path))
             print('Done computing file size')
-        helpers.batch(process_rows, rows, chunksize=100, total=total)
+        all(helpers.batch(process_rows, rows, chunksize=100, total=total))
```
Comment (davidbgk) [inline ban/commands/init.py]: I guess that one is required now that we `yield from results`? Is that the less expensive available evaluation? If yes, document? (If no, do document 😛)

Comment (yohanboniface) [inline ban/commands/init.py]: Done :)

File: ban/db/cache.py (hunk 1 above)
Comment (yohanboniface) [inline ban/db/cache.py]: Fixed :)

File: ban/db/fields.py (hunk 2 above)
Comment (yohanboniface) [inline ban/db/fields.py]: Humm, not sure what info would be helpful here :s
You have something in mind?

File: ban/commands/export.py (hunk 5 above)
Comment (yohanboniface) [inline ban/commands/export.py]: Given they are only used once, I didn't feel the need. You'd prefer to have some?

File: ban/commands/export.py (hunk 3 above)
Comment (yohanboniface) [inline ban/commands/export.py]: Yeah, I'm not sure what we want to do here.
Basically:
- `f.write` will write to the python buffer
- `f.flush` will transfer the data to the OS buffers (I guess this is where we release memory)
- `os.fsync` will actually write on disk

So I added the `os.fsync` line to be able to live control the size of the file for debug purposes, but I think it's better to actually let the OS decide when to write on disk (we don't really care about the scenario where the server lose power during the export process — in this case we will just rerun the export anyway), so maybe it's better to remove `os.fsync`.
But I'd keep the `flush` call, as I understand this is where python will release memory.

Thoughts?

File: ban/commands/export.py (hunk 4 above)
Comment (yohanboniface) [inline ban/commands/export.py]: Done :)

File: ban/commands/export.py (hunk 5 above)
Comment (davidbgk) [inline ban/commands/export.py]: I was wondering if it affects performances and if you have to modify these based on your “hardware” config?

File: ban/commands/export.py (hunk 3 above)
Comment (davidbgk) [inline ban/commands/export.py]: I would remove the `fsync` and test if the flush actually release some memory or not. Sometimes Python is more clever than we think :wink:

File: ban/db/fields.py (hunk 2 above)
Comment (davidbgk) [inline ban/db/fields.py]: Idea: having an explicit `CachedForeignKey`?

**Comment by yohanboniface** [comment]
> Plusieurs lignes des fichiers exemples produits avec ce code ne sont pas valides.
En fait le séparateur de ligne ne semble pas systématiquement ajouté.

Oui, j'ai corrigé ce problème, j'aurais dû te refaire un export, sorry! Je m'occupe de ça demain.

**Comment by jdesboeufs** [comment]
Super :)
Du coup cette PR est à jour ? Elle inclut la correction ?

**Comment by yohanboniface** [comment]
>  Du coup cette PR est à jour ? Elle inclut la correction ?

Yep

File: ban/commands/export.py (hunk 3 above)
Comment (yohanboniface) [inline ban/commands/export.py]: Done!

File: ban/db/fields.py (hunk 2 above)
Comment (yohanboniface) [inline ban/db/fields.py]: Nice suggestion, I've changed to that implementation :)

**Pull request merged**